programs/library-bridge/SharedLibraryHandler.cpp

```diff
@@ -0,0 +1,233 @@
+
+        if (lib_new)
+            lib_data = lib_new(&settings_holder->strings, ClickHouseLibrary::log);
+    }
+}
+
+
+SharedLibraryHandler::~SharedLibraryHandler()
+{
+    auto lib_delete = library->tryGet<ClickHouseLibrary::LibraryDeleteFunc>(ClickHouseLibrary::LIBRARY_DELETE_FUNC_NAME);
+
+    if (lib_delete)
+        lib_delete(lib_data);
+}
+
+
+bool SharedLibraryHandler::isModified()
+{
+    auto func_is_modified = library->tryGet<ClickHouseLibrary::LibraryIsModifiedFunc>(ClickHouseLibrary::LIBRARY_IS_MODIFIED_FUNC_NAME);
+
+    if (func_is_modified)
+        return func_is_modified(lib_data, &settings_holder->strings);
+
+    return true;
+}
+
+
+bool SharedLibraryHandler::supportsSelectiveLoad()
+{
+    auto func_supports_selective_load = library->tryGet<ClickHouseLibrary::LibrarySupportsSelectiveLoadFunc>(ClickHouseLibrary::LIBRARY_SUPPORTS_SELECTIVE_LOAD_FUNC_NAME);
+
+    if (func_supports_selective_load)
+        return func_supports_selective_load(lib_data, &settings_holder->strings);
+
+    return true;
+}
+
+
+BlockInputStreamPtr SharedLibraryHandler::loadAll(const std::string & attributes_string, const Block & sample_block)
+{
+    std::vector<std::string> dict_attributes;
+    boost::split(dict_attributes, attributes_string, [](char c) { return c == ' '; });
+
+    auto columns_holder = std::make_unique<ClickHouseLibrary::CString[]>(dict_attributes.size());
+    ClickHouseLibrary::CStrings columns{static_cast<decltype(ClickHouseLibrary::CStrings::data)>(columns_holder.get()), dict_attributes.size()};
+    size_t i = 0;
+
+    for (const auto & attr : dict_attributes)
+        columns.data[i++] = attr.c_str();
+
+    auto load_all_func = library->get<ClickHouseLibrary::LibraryLoadAllFunc>(ClickHouseLibrary::LIBRARY_LOAD_ALL_FUNC_NAME);
+    auto data_new_func = library->get<ClickHouseLibrary::LibraryDataNewFunc>(ClickHouseLibrary::LIBRARY_DATA_NEW_FUNC_NAME);
+    auto data_delete_func = library->get<ClickHouseLibrary::LibraryDataDeleteFunc>(ClickHouseLibrary::LIBRARY_DATA_DELETE_FUNC_NAME);
+
+    ClickHouseLibrary::LibraryData data_ptr = data_new_func(lib_data);
+    SCOPE_EXIT(data_delete_func(lib_data, data_ptr));
+
+    ClickHouseLibrary::RawClickHouseLibraryTable data = load_all_func(data_ptr, &settings_holder->strings, &columns);
+    auto block = dataToBlock(sample_block, data);
+
+    return std::make_shared<OneBlockInputStream>(block);
+}
+
+
+BlockInputStreamPtr SharedLibraryHandler::loadIds(const std::string & attributes_string, const std::string & ids_string, const Block & sample_block)
+{
+    std::vector<std::string> dict_string_ids;
+    boost::split(dict_string_ids, ids_string, [](char c) { return c == ' '; });
+    std::vector<UInt64> dict_ids(dict_string_ids.size());
+    for (const auto & id : dict_string_ids)
+        dict_ids.push_back(parseFromString<UInt64>(id));
+
+    std::vector<std::string> dict_attributes;
```

Comment:
We serialize attributes to a string, passing to bridge, deserialize them just to take size?!